controllers/clustermesh/clustermesh_controller.go

```diff
@@ -74,117 +86,221 @@ type ClusterMeshReconciler struct {
-	start := time.Now()
-	r.log = ctrl.LoggerFrom(ctx)
+func DefaultReconciler(mgr manager.Manager) *ClusterMeshReconciler {
+	r := &ClusterMeshReconciler{
+		Client:                         mgr.GetClient(),
+		Scheme:                         mgr.GetScheme(),
+		NewEC2ClientFactory:            ec2.NewEC2Client,
+		PopulateClusterSpecFactory:     PopulateClusterSpec,
+		ReconcilePeeringsFactory:       ReconcilePeerings,
+		ReconcileSecurityGroupsFactory: ReconcileSecurityGroups,
+		ReconcileRoutesFactory:         ReconcileRoutes,
+	}
+
+	return r
+}
+
+func awsConfigForCredential(ctx context.Context, region string, accessKey string, secretAccessKey string) (aws.Config, error) {
+	return config.LoadDefaultConfig(ctx,
+		config.WithRegion(region),
+		config.WithCredentialsProvider(credentials.StaticCredentialsProvider{
+			Value: aws.Credentials{
+				AccessKeyID:     accessKey,
+				SecretAccessKey: secretAccessKey,
+			},
+		}),
+	)
+}
+
+func (c *ClusterMeshReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r := &ClusterMeshReconciliation{
+		ClusterMeshReconciler: *c,
+		start:                 time.Now(),
+		log:                   ctrl.LoggerFrom(ctx),
+	}
 
 	cluster := &clusterv1beta1.Cluster{}
-	if err := r.Get(ctx, req.NamespacedName, cluster); err != nil {
+	if err := c.Get(ctx, req.NamespacedName, cluster); err != nil {
 		return resultError, err
 	}
 
-	clustermesh := &clustermeshv1beta1.ClusterMesh{}
+	if cluster.Spec.ControlPlaneRef == nil || cluster.Spec.ControlPlaneRef.Kind != "KopsControlPlane" {
+		return resultDefault, fmt.Errorf("controlPlaneRef not specified or not supported")
+	}
+
+	r.kcp = &kcontrolplanev1alpha1.KopsControlPlane{}
+	key := client.ObjectKey{
+		Namespace: cluster.Spec.ControlPlaneRef.Namespace,
+		Name:      cluster.Spec.ControlPlaneRef.Name,
+	}
+	if err := c.Get(ctx, key, r.kcp); err != nil {
+		return requeue1min, err
+	}
+	subnet, err := kops.GetSubnetFromKopsControlPlane(r.kcp)
+	if err != nil {
+		r.log.Error(err, "failed to get subnet from kcp")
+		return requeue1min, err
+	}
+
+	region, err := kops.GetRegionFromKopsSubnet(*subnet)
+	if err != nil {
+		r.log.Error(err, "failed to get region from kcp")
+		return requeue1min, err
+	}
 
+	var secretName, namespace string
+	if r.kcp.Spec.IdentityRef != nil {
+		secretName = r.kcp.Spec.IdentityRef.Name
+		// gambiarra??
+		if r.kcp.Spec.IdentityRef.Namespace == "" {
+			namespace = "kubernetes-kops-operator-system"
+		} else {
+			namespace = r.kcp.Spec.IdentityRef.Namespace
+		}
+	} else {
+		secretName = "default"
+		namespace = "kubernetes-kops-operator-system"
+	}
+
+	r.providerConfigName = secretName
+
+	awsCreds := &corev1.Secret{}
+	key = client.ObjectKey{
+		Namespace: namespace,
+		Name:      secretName,
+	}
+	if err := c.Get(ctx, key, awsCreds); err != nil {
+		r.log.Error(err, "failed to get secret")
+		return requeue1min, err
+	}
+
+	accessKeyBytes, ok := awsCreds.Data["AccessKeyID"]
```

Comment:
WDYT about encapsulating this logic part of retrieving the secret? I think it would improve readability and make easier to create the test cases